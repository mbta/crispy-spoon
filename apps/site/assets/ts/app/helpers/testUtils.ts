/* eslint-disable */
import toJson, { Json } from "enzyme-to-json";
import { ReactWrapper } from "enzyme";
import {
  IndexedRouteStop,
  StopTree
} from "../../schedule/components/__schedule";

export const testRouteStopListFromStopTree = (
  tree: StopTree
): IndexedRouteStop[] =>
  Object.values(tree.byId).map((node, index) => ({
    ...node.value,
    routeIndex: index
  }));

export const createReactRoot = (): void => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';
};

export const enzymeToJsonWithoutProps = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false,
    map: info => {
      if (!info.props) {
        return info;
      }

      // @ts-ignore is a <style></style> declaration
      if (info.node.type === "style") {
        return info;
      }
      // @ts-ignore is an element in a <noscript></noscript> declaration
      if (info.node.nodeType == undefined) {
        // @ts-ignore
        return info;
      }
      //@ts-ignore is a React component (not HTML element)
      if (info.node.nodeType.toString().includes("function")) {
        // @ts-ignore
        info.props = undefined;
      }

      return info;
    }
  });

export const toJsonEnzyme = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false
  });
