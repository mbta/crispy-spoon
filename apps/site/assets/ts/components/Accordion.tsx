import React, { ReactElement, useState } from "react";

// A simple accordion implementation, based on site_web/templates/partial/_accordion_ui
// This attempts to mimic the usage of <ExpandableBlock />, where the contents rerender with expanded state
interface Props {
  id: string;
  title: {
    collapsed: string;
    expanded: string;
  };
  children: ReactElement<HTMLElement>;
}

export const AccordionNoJS = ({
  id,
  children
}: {
  id: string;
  children: ReactElement<HTMLElement>;
}): ReactElement<HTMLElement> => (
  <div className="c-accordion-ui">
    <div
      className="c-accordion-ui__target"
      id={`${id}-section`}
      role="region"
      aria-labelledby={`${id}-title`}
    >
      <div className="c-accordion-ui__content">{children}</div>
    </div>
  </div>
);

const Accordion = (props: Props): ReactElement<HTMLElement> => {
  const { id, title, children } = props;
  const [isExpanded, toggleExpanded] = useState(false);
  const onClick = (): void => toggleExpanded(expanded => !expanded);
  return (
    <div className="c-accordion-ui">
      <div className="panel">
        <div className="c-accordion-ui__heading" role="heading" aria-level={3}>
          <button
            className={`c-accordion-ui__trigger ${!isExpanded && "collapsed"}`}
            aria-expanded={isExpanded}
            aria-controls={`${id}-section`}
            onClick={onClick}
            type="button"
          >
            <span className="c-accordion-ui__title" id={`${id}-title`}>
              {isExpanded ? title.expanded : title.collapsed}
            </span>
            <div className="c-accordion-ui__indicator">
              <span className="c-indicator__content c-indicator__content--angle" />
            </div>
          </button>
        </div>
        <div
          className="c-accordion-ui__target js-focus-on-expand"
          id={`${id}-section`}
          role="region"
          aria-labelledby={`${id}-title`}
        >
          {isExpanded && (
            <div className="c-accordion-ui__content">{children}</div>
          )}
        </div>
      </div>
    </div>
  );
};

export default Accordion;
