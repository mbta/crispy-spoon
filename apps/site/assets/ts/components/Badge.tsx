import React from "react";

const Badge = ({
  text,
  contextText,
  bgClass
}: {
  text: string;
  contextText?: string;
  bgClass?: string;
}): JSX.Element => {
  return (
    <div
      className={`${bgClass ||
        "u-error-background"} font-weight-bold ps-8 pe-8 fs-14`}
      style={{ borderRadius: "0.75rem" }}
    >
      {/* The purpose of this block is to have invisble text for screen readers */}
      {contextText && <span className="sr-only">{contextText}</span>}
      {text}
    </div>
  );
};

export default Badge;
