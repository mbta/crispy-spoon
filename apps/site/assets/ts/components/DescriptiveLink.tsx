import React from "react";
import renderFa from "../helpers/render-fa";

interface DescriptiveLinkProps {
  href: string;
  title: string;
  bodyTextOrHtml: string; // might be HTML from the CMS
}

/* Since the props can be populated by running CMS content through
`Site.ContentRewriter.rewrite` this might end up using an empty string in place
of null values. So we can't do our usual undefined/null checks. This helper
function will substitute. */
const hasContent = (text: string): boolean => text !== "";

const DescriptiveLink = ({
  href,
  title,
  bodyTextOrHtml
}: DescriptiveLinkProps): JSX.Element | null => {
  const titleContent = hasContent(title) && (
    <div className="c-descriptive-link__title">{title}</div>
  );
  const bodyContent = hasContent(bodyTextOrHtml) && (
    // eslint-disable-next-line react/no-danger
    <div dangerouslySetInnerHTML={{ __html: bodyTextOrHtml }} />
  );

  if (!titleContent && !bodyContent) {
    return null;
  }

  return (
    <a className="c-descriptive-link" href={href}>
      <div className="c-descriptive-link__text">
        {titleContent}
        {bodyContent}
      </div>
      <div className="c-descriptive-link__caret-wrapper">
        {renderFa("c-descriptive-link__caret", "fa-solid fa-angle-right")}
      </div>
    </a>
  );
};

export default DescriptiveLink;
