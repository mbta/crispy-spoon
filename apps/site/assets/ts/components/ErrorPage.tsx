import React from "react";
import { FallbackRender } from "@sentry/react";

// Adapted from the templates in SiteWeb.ErrorView
// To be used in an ErrorBoundary
const ErrorPage: FallbackRender = errorData => {
  // eslint-disable-next-line no-console
  console.error("from Sentry", JSON.stringify(errorData));
  return (
    <div className="container error-page">
      <div className="row">
        <div className="col-md-6">
          <h1 className="mt-1">Oh no! We&#39;re experiencing delays.</h1>
          <p className="error-paragraph u-bold">
            Something went wrong on our end.
          </p>
          <div className="error-links">
            Go to{" "}
            <a className="c-call-to-action" href="/">
              {`the MBTA's home page`}
            </a>{" "}
            or{" "}
            <a className="c-call-to-action" href="/customer-support">
              report an issue
            </a>
            .
          </div>
        </div>
        <div className="col-md-6">
          <img alt="" src="/images/error-bus.gif" className="w-100" />
        </div>
      </div>
    </div>
  );
};

export default ErrorPage;
