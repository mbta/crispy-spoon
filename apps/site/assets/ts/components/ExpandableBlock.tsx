import React, { ReactElement, useState } from "react";
import { isString } from "lodash";
import { handleReactEnterKeyPress } from "../helpers/keyboard-events-react";
import renderSvg from "../helpers/render-svg";
import renderFa from "../helpers/render-fa";
import { caret } from "../helpers/icon";

export interface ExpandableBlockHeader {
  text: string | ReactElement<HTMLElement>;
  iconSvgText: string | null;
  classOverride?: string;
}

// If dispatch is provided in Props, the block will not
// track its own state -- the parent is fully responsible
// for tracking the expanded/collapsed state.
// If dispatch is not provided, the block will
// track its own state.
interface Props {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  dispatch?: React.Dispatch<any>;
  notifyExpanded?: (isExpanded: boolean) => void;
  initiallyExpanded: boolean;
  initiallyFocused?: boolean;
  header: ExpandableBlockHeader;
  children: ReactElement<HTMLElement>;
  id: string;
  panelClassName?: string;
  preventScroll?: boolean;
}

interface State {
  expanded: boolean;
  focused?: boolean;
}

export interface ClickExpandableBlockAction {
  type: "CLICK_EXPANDABLE_BLOCK";
  payload: {
    expanded: boolean;
    focused: boolean;
    id: string;
  };
}

const ExpandableBlock = (props: Props): ReactElement<HTMLElement> => {
  const {
    header: { text, iconSvgText, classOverride },
    dispatch,
    notifyExpanded,
    initiallyExpanded,
    initiallyFocused,
    children,
    id,
    panelClassName,
    preventScroll
  } = props;

  const initialState = {
    expanded: initiallyExpanded,
    focused: initiallyFocused
  };

  const action: ClickExpandableBlockAction = {
    type: "CLICK_EXPANDABLE_BLOCK",
    payload: {
      expanded: initiallyExpanded === true,
      focused: initiallyFocused === true,
      id
    }
  };

  const [hookedState, toggleExpanded] = useState(initialState);
  const { state, onClick } = dispatch
    ? {
        state: initialState,
        onClick: () => dispatch(action)
      }
    : {
        state: hookedState,
        onClick: () => {
          const expanded = !hookedState.expanded;
          toggleExpanded({ expanded, focused: true });
          if (notifyExpanded) notifyExpanded(expanded);
        }
      };
  const { expanded, focused }: State = state;
  const headerId = `header-${id}`;
  const panelId = `panel-${id}`;
  const iconHelper = iconSvgText?.slice(0, 3) !== "fa-" ? renderSvg : renderFa;
  const classOverrideString = isString(classOverride) ? classOverride : "";

  return (
    <>
      <h3
        className={`c-expandable-block__header ${classOverrideString}`}
        tabIndex={0}
        id={headerId}
        aria-expanded={expanded}
        aria-controls={panelId}
        // eslint-disable-next-line jsx-a11y/no-noninteractive-element-to-interactive-role
        role="button"
        onClick={onClick}
        onKeyPress={e => handleReactEnterKeyPress(e, onClick)}
      >
        {iconSvgText
          ? iconHelper("c-expandable-block__header-icon", iconSvgText)
          : null}
        <div className="c-expandable-block__header-text">
          {text}
          {caret("c-expandable-block__header-caret", expanded)}
        </div>
      </h3>
      {expanded ? (
        <div
          className={`c-expandable-block__panel ${panelClassName || ""}`}
          // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
          tabIndex={0}
          role="region"
          id={panelId}
          aria-labelledby={headerId}
          ref={panel =>
            panel &&
            focused &&
            panel.focus({
              preventScroll: !!preventScroll
            })
          }
        >
          {children}
        </div>
      ) : null}
      {/* No javascript support */}
      <noscript>
        <style>{`#${headerId} { display: none; }`}</style>
        <h3 className="c-expandable-block__header">
          {iconSvgText
            ? renderSvg("c-expandable-block__header-icon", iconSvgText)
            : null}
          {text}
          {caret("c-expandable-block__header-caret", true)}
        </h3>
        <div className="c-expandable-block__panel" role="region">
          {children}
        </div>
      </noscript>
    </>
  );
};

export default ExpandableBlock;
