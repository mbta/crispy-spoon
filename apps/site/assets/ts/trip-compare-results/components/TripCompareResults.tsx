import React, { ReactElement, useState } from "react";
import { handleReactEnterKeyPress } from "../../helpers/keyboard-events";
import TripPlannerResults, { Itinerary } from "./TripPlannerResults";

interface Props {
  // eslint-disable-next-line
  itineraryData: Itinerary[];
  itineraryHeader: string;
}

const TripCompareResults = ({
  itineraryData,
  itineraryHeader
}: Props): ReactElement<HTMLElement> => {
  const [source, setSource] = useState("NEW");

  const filteredData = itineraryData.filter(i => i.source === source);
  const classRedesigned =
    source !== "NEW"
      ? "m-alerts__mode-button"
      : "m-alerts__mode-button m-alerts__mode-button--selected";
  const classCurrent =
    source === "NEW"
      ? "m-alerts__mode-button"
      : "m-alerts__mode-button m-alerts__mode-button--selected";
  const note =
    source === "NEW"
      ? "The following trips only reflect normal service within our proposed redesigned bus network. They do not reflect current bus service and do not include bus service alerts. To learn more about the fares associated with your trip, please visit https://www.mbta.com/fares."
      : "The following trips only reflect normal service (i.e. do not reflect current service, do not include service alert info)";

  const onClickNew = (): void => setSource("NEW");
  const onClickCurrent = (): void => setSource("CURRENT");

  return (
    <>
      <h2>Compare trips</h2>
      <div>
        <div className="m-alerts__mode-buttons">
          <div className="m-alerts__mode-button-container">
            <div
              className={classRedesigned}
              onClick={onClickNew}
              onKeyPress={e => handleReactEnterKeyPress(e, onClickNew)}
              role="button"
              tabIndex={0}
            >
              <div className="m-alerts__mode-button-name">Redesigned</div>
            </div>
          </div>
          <div className="m-alerts__mode-button-container">
            <div
              className={classCurrent}
              onClick={onClickCurrent}
              onKeyPress={e => handleReactEnterKeyPress(e, onClickCurrent)}
              role="button"
              tabIndex={0}
            >
              <div className="m-alerts__mode-button-name">Current</div>
            </div>
          </div>
        </div>
        <p className="no-trips page-section">
          We found {filteredData.length} trips for you
        </p>
        <p className="instructions page-section">{itineraryHeader}</p>
        <p className="instructions page-section">
          <b>Note: </b>
          {note}
        </p>
        <TripPlannerResults itineraryData={filteredData} />
      </div>
    </>
  );
};

export default TripCompareResults;
