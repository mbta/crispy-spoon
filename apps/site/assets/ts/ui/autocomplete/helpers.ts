import { omit } from "lodash";
import { OnSubmitParams } from "@algolia/autocomplete-core";
import { OnStateChangeProps } from "@algolia/autocomplete-js";
import {
  ContentItem,
  Item,
  LocationItem,
  PopularItem,
  RouteItem,
  StopItem
} from "./__autocomplete";
import { isLGDown } from "../../helpers/media-breakpoints";

export function isStopItem(x: Item): x is StopItem {
  return Object.keys(x).includes("stop");
}

export function isRouteItem(x: Item): x is RouteItem {
  return Object.keys(x).includes("route");
}

export function isContentItem(x: Item): x is ContentItem {
  return Object.keys(x).includes("_content_type");
}

export const getTitleAttribute = (item: Item): string[] => {
  if (isStopItem(item)) {
    return ["stop", "name"];
  }
  if (isRouteItem(item)) {
    return ["route", "name"];
  }

  // @ts-ignore
  // eslint-disable-next-line no-underscore-dangle
  return Object.keys(item._highlightResult);
};

const navStateChange: (props: OnStateChangeProps<Item>) => void = ({
  state
}) => {
  // grey out the page and disable scrolling when search is open
  if (isLGDown()) {
    if (state.isOpen) {
      document.documentElement.dataset.navOpen = "true";
    } else {
      delete document.documentElement.dataset.navOpen;
    }
  }
};

export const STATE_CHANGE_HANDLERS: Record<
  string,
  (props: OnStateChangeProps<Item>) => void
> = {
  nav: navStateChange
};

export const SUBMIT_HANDLERS: Record<
  string,
  (props: OnSubmitParams<Item>) => void
> = {
  to_search_page({ state }) {
    window.Turbolinks.visit(`/search?query=${state.query}`);
  }
};

export type WithUrls<T> = T & { urls: Record<string, string> };

// The backend returns all possible URLs, use urlType to get the desired one
export const itemWithUrl = (
  initialItem: WithUrls<LocationItem | PopularItem>,
  urlType: string
): Omit<typeof initialItem, "urls"> => {
  const item = omit(initialItem, "urls");
  item.url = initialItem.urls[urlType];
  return item;
};
