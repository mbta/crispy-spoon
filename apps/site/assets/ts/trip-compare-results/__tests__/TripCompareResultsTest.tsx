import React from "react";
import renderer, { act } from "react-test-renderer";
import { createReactRoot } from "../../app/helpers/testUtils";
import TripCompareResults from "../components/TripCompareResults";
import Accordion, { AccordionNoJS } from "../../components/Accordion";
import { TileServerUrl } from "../../leaflet/components/__mapdata";

const html = "<div>Lots of content about the itinerary</div>";
const tab_html = "<div>HTML for the accordion button</div>";
const access_html = "<div>Accessible</div>";
const fares_estimate_html = "<div>HTML content for fares</div>";
const fare_calculator_html = "<div>HTML content for fare calculator</div>";
const tile_server_url: TileServerUrl = "";

const positions: [number, number][] = [
  [42.3564, -71.06242],
  [42.35628, -71.06226]
];
const map = {
  zoom: null,
  markers: [],
  tile_server_url,
  default_center: { longitude: 0, latitude: 0 },
  height: 400,
  width: 400,
  polylines: [
    {
      color: "#428608",
      "dotted?": false,
      id: "",
      positions,
      weight: 5
    }
  ]
};

it("it renders", () => {
  createReactRoot();
  const tree = renderer.create(
    <TripCompareResults
      itineraryData={[
        {
          html,
          access_html,
          fares_estimate_html,
          tab_html,
          fare_calculator_html,
          id: 1,
          source: "NEW",
          map
        }
      ]}
      itineraryHeader="header"
    />
  );
  tree.update(
    <TripCompareResults
      itineraryData={[
        {
          html,
          access_html,
          fares_estimate_html,
          tab_html,
          fare_calculator_html,
          id: 1,
          source: "NEW",
          map
        }
      ]}
      itineraryHeader="header"
    />
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

it("it renders ItineraryBody when clicking to expand", () => {
  createReactRoot();
  const tree = renderer.create(
    <TripCompareResults
      itineraryData={[
        {
          html,
          access_html,
          fares_estimate_html,
          tab_html,
          fare_calculator_html,
          id: 1,
          source: "NEW",
          map
        }
      ]}
      itineraryHeader="header"
    />
  );

  const button = tree.root.findAllByType("button")[0];
  act(button.props.onClick);

  expect(tree.root.findAllByType(Accordion).length).toBe(1);
  expect(tree.root.findAllByType(AccordionNoJS).length).toBe(1);

  expect(
    tree.root.findAllByProps({
      html: "<div>Lots of content about the itinerary</div>"
    }).length
  ).toBe(2);
});

it("it switches between current and redesigned modes", () => {
  createReactRoot();
  const tree = renderer.create(
    <TripCompareResults
      itineraryData={[
        {
          html,
          access_html,
          fares_estimate_html,
          tab_html,
          fare_calculator_html,
          id: 1,
          source: "NEW",
          map
        }
      ]}
      itineraryHeader="header"
    />
  );

  const button = tree.root.findByProps({
    className: "m-alerts__mode-button"
  });
  act(button.props.onClick);

  expect(tree.root.findAllByType(Accordion).length).toBe(0);
  expect(tree.root.findAllByType(AccordionNoJS).length).toBe(0);

  expect(
    tree.root.findAllByProps({
      html: "<div>Lots of content about the itinerary</div>"
    }).length
  ).toBe(0);
});
