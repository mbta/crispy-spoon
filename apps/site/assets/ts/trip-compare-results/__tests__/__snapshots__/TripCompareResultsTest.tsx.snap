// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
Array [
  <h2>
    Compare trips
  </h2>,
  <div>
    <div
      className="m-alerts__mode-buttons"
    >
      <div
        className="m-alerts__mode-button-container"
      >
        <div
          className="m-alerts__mode-button m-alerts__mode-button--selected"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="m-alerts__mode-button-name"
          >
            Redesigned
          </div>
        </div>
      </div>
      <div
        className="m-alerts__mode-button-container"
      >
        <div
          className="m-alerts__mode-button"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="m-alerts__mode-button-name"
          >
            Current
          </div>
        </div>
      </div>
    </div>
    <p
      className="no-trips page-section"
    >
      We found 
      1
       trips for you
    </p>
    <p
      className="instructions page-section"
    >
      header
    </p>
    <p
      className="instructions page-section"
    >
      <b>
        Note: 
      </b>
      The following trips only reflect normal service within our proposed redesigned bus network. They do not reflect current bus service and do not include bus service alerts
    </p>
    <div
      className="m-trip-plan-results__itinerary"
    >
      <div
        className="m-trip-plan-results__itinerary-header"
      >
        <div
          className="m-trip-plan-results__itinerary-header-content"
        >
          <div
            className="m-trip-plan-results__itinerary-summary"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<div>HTML for the accordion button</div>",
              }
            }
          />
        </div>
      </div>
      <div
        className="c-accordion-ui"
      >
        <div
          aria-level={3}
          className="panel"
          role="heading"
        >
          <div
            className="c-accordion-ui__heading"
          >
            <button
              aria-controls="itinerary-1-section"
              aria-expanded={false}
              className="c-accordion-ui__trigger collapsed"
              onClick={[Function]}
              type="button"
            >
              <span
                className="c-accordion-ui__title"
                id="itinerary-1-title"
              >
                Show map and trip details
              </span>
              <div
                className="c-accordion-ui__indicator"
              >
                <span
                  className="c-indicator__content c-indicator__content--angle"
                />
              </div>
            </button>
          </div>
          <div
            aria-labelledby="itinerary-1-title"
            className="c-accordion-ui__target js-focus-on-expand"
            id="itinerary-1-section"
            role="region"
          />
        </div>
      </div>
      <noscript>
        <div
          className="c-accordion-ui"
        >
          <div
            aria-level={3}
            role="heading"
          >
            <div
              aria-labelledby="itinerary-1-title"
              className="c-accordion-ui__target"
              id="itinerary-1-section"
              role="region"
            >
              <div
                className="c-accordion-ui__content"
              >
                <div
                  className="m-trip-plan-results__itinerary-body"
                >
                  <div
                    className="trip-plan-map map"
                  />
                  <div
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "<div>Lots of content about the itinerary</div>",
                      }
                    }
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </noscript>
    </div>
  </div>,
]
`;
